packages/langium/src/workspace/document-builder.ts

```diff
@@ -251,21 +251,28 @@ export class DefaultDocumentBuilder implements DocumentBuilder {
      * in files that are currently not opened in the editor.
      */
     protected sortDocuments(documents: LangiumDocument[]): LangiumDocument[] {
-        const hasTextDocument = new Map<LangiumDocument, boolean>();
-        for (const doc of documents) {
-            hasTextDocument.set(doc, Boolean(this.textDocuments?.get(doc.uri.toString())));
-        }
-        return documents.sort((a, b) => {
-            const aHasDoc = hasTextDocument.get(a);
-            const bHasDoc = hasTextDocument.get(b);
-            if (aHasDoc && !bHasDoc) {
-                return -1;
-            } else if (!aHasDoc && bHasDoc) {
-                return 1;
-            } else {
-                return 0;
+        let left = 0;
+        let right = documents.length - 1;
+
+        while (left < right) {
+            while (left < documents.length && this.hasTextDocument(documents[left])) {
+                left++;
             }
-        });
+
+            while (right >= 0 && !this.hasTextDocument(documents[right])) {
+                right--;
+            }
+
+            if (left < right) {
+                [documents[left], documents[right]] = [documents[right], documents[left]];
+            }
+        }
```

Comment:
Could you provide a unit test for this algorithm by mocking the `textDocuments` service?

Reply:
Sure - is a naive method extraction acceptable? If not, it will take me a minute to figure out how to wield the dark arts of dependency injection to conjure up a modified `createServicesForGrammar()` function (if that's even feasible).

Reply:
IMO it's ok to override TypeScript's `protected` method visibility by casting the service to `any` in the test.

Reply:
Thanks for the hint - test code isn't perfect, but I think it shows that it works.